### PR TITLE
Add support for the Firefox screen recorder to calculate visual metrics

### PR DESCRIPTION
**Pull request by dpalmeiro**
[bug 1536174](https://bugzilla.mozilla.org/show_bug.cgi?id=1536174) now gives us the ability to save screenshots to disk whenever a meaningful change occurs during composition.  This PR will utilize this new feature to calculate visual metrics.

### CODE REVIEW

File: lib/core/engine/collector.js [hunk 1]
```diff
@@ -168,8 +168,8 @@ class Collector {
               data.visualMetrics.FirstVisualChange
             } SpeedIndex: ${
               data.visualMetrics.SpeedIndex
-            } PerceptualSpeedIndex: ${
```
Comment (ncalexan) [inline lib/core/engine/collector.js]: This should be a separate commit.  Is upstream likely to take this change?  Would they take just _adding_ `VisualComplete85`?

File: lib/core/engine/command/measure.js [hunk 2]
```diff
@@ -61,9 +61,18 @@ class Measure {
 
   // Have a concistent way of starting the video
   async _startVideo(numberOfMeasuredPages, index) {
-    this.video = new Video(this.storageManager, this.options);
-    await setOrangeBackground(this.browser.getDriver(), this.options);
-    await this.video.record(numberOfMeasuredPages, index);
+    this.video = new Video(this.storageManager, this.options, this.browser);
+
+    if (this.options.firefox.windowRecorder) {
```
Comment (ncalexan) [inline lib/core/engine/command/measure.js:66]: Awkward.  Any reason to not set orange before _and_ after we start recording?  IIUC, the existing code will look for orange... is it possible that an orange-to-orange transition will look like a first paint to the existing system?

File: lib/support/cli.js [hunk 3]
```diff
@@ -231,6 +231,12 @@ module.exports.parseCommandLine = function parseCommandLine() {
       type: 'boolean',
       group: 'firefox'
     })
+    .option('firefox.windowRecorder', {
+      describe: 'Use the Firefox internal window recorder',
```
Comment (ncalexan) [inline lib/support/cli.js]: Please include something about "video capture" here.  Should this be in the `video.*` section?

File: lib/support/cli.js [hunk 4]
```diff
@@ -231,6 +231,12 @@ module.exports.parseCommandLine = function parseCommandLine() {
...
+      default: true,
```
Comment (ncalexan) [inline lib/support/cli.js:239]: Really?

File: lib/video/postprocessing/visualmetrics/single/visualMetrics.js [hunk 5]
```diff
@@ -9,7 +9,7 @@ const SCRIPT_PATH = require('../visualMetricsPath');
 
 module.exports = {
   async checkDependencies() {
-    return execa(SCRIPT_PATH, ['--check']);
+    return execa('python', [SCRIPT_PATH, '--check']);
```
Comment (ncalexan) [inline lib/video/postprocessing/visualmetrics/single/visualMetrics.js:12]: Break this into a separate changeset, and fold the `--perceptual` piece into the earlier commit.  We want to get this upstream, right?

Also, how about looking for `$PYTHON` in the environment, and then falling back to `python` on the `$PATH`?

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
+'use strict';
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: License block?  `firefoxWindowRecorder.js` with capitalization?  (I hate that `browsertime` doesn't name files, types, and imports exactly the same -- it makes grepping unnecessarily difficult.)

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+const unlink = util.promisify(fs.unlink);
+
+async function createTempFile() {
+  const mkdtemp = util.promisify(fs.mkdtemp);
+  const dir = await mkdtemp(path.join(os.tmpdir(), 'browsertime-'));
+  return path.join(dir, 'winrecorder.mp4');
+}
+
+async function enableWindowRecorder(enable, driver) {
+  if (enable) {
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: I'd prefer to use my privileged JS bits, but no trouble: just `// TODO: ...` something here as a reminder.

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  const script = 'windowUtils.setCompositionRecording(' + enable + ');';
+  await driver.executeScript(script);
+  return driver.setContext(oldContext);
+}
+
+function findRecordingDirectory() {
+  let closest_mtime = 0;
+  let directory = undefined;
+
+  fs.readdirSync('.').forEach(file => {
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: What are you trying to find?  This doesn't find the "smallest greater than", it finds the "greatest greater than".  Maybe collect all `(mtime, file)`, sort, filter, take the head?  Can we not just save something that will allow us to determine this file number?  Maybe we need to make Firefox tell us where it's writing the files?

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      }
+    }
+  });
+
+  log.debug('Using window recording directory: ' + directory);
+  return directory;
+}
+
+async function pollDirectory(directory) {
+  // We need to spin wait here until the browser is finished
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: Add a `// TODO: ...` saying the bug that will make the actual API in Firefox useful (synchronous).

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        new_mtime.getTime()
+    );
+    await delay(5000);
+    old_mtime = new_mtime;
+    new_mtime = fs.statSync(directory).mtime;
+  }
+}
+
+async function generateVideo(destination) {
+  let directoryName = findRecordingDirectory();
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: Is it possible that this races Firefox to write even a single image, meaning we'll fail entirely?

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  await pollDirectory(directoryName);
+  let imageFiles = [];
+
+  fs.readdirSync(directoryName).forEach(file => {
+    // Format of the filenames are "frame-<num>-<offset>.png"
+    // where num is frame number and offset is time since capture start in ms.
+    let fields = file.split('-');
+    let frameno = ('0000' + fields[1]).slice(-4);
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: Are you telling me that `num` is unjustified, i.e., runs `1, ..., 10, ..., 100`?  That's extremely unhelpful :(

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    let newFilename = 'frame' + frameno + '.png';
+    let offset = fields[2].split('.')[0];
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: I don't care too much right now, but this would be simpler if you stripped the extension, split the fields, and then converted to integers.  Then you have more control over justification (and error cases!) when you recombine.

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    imageFiles.push({ filename: newFilename, offset: offset });
+  });
+
+  imageFiles.sort(function(a, b) {
+    if (a.filename < b.filename) return -1;
+    if (a.filename > b.filename) return 1;
+    return 0;
+  });
+
+  //First step is merging all of the png frames into a single CFR video.
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: nit: `// First ...` (space).  Expand `constant frame rate/CFR`, please.

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    '-i',
+    path.join(directoryName,'frame%04d.png'),
+    '-vf',
+    'pad=ceil(iw/2)*2:ceil(ih/2)*2',
+    '-pix_fmt',
+    'yuv420p',
+    path.join(directoryName,'tmp-cfr.mp4')
+  ];
+  log.debug('Executing command: ffmpeg ' + cfr_args.join(' '));
+  await execa('ffmpeg', cfr_args);
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: Wish we weren't taking `ffmpeg` from the path, but we'll clean this up at some point in the future.  (Gotta do it for `visualmetrics.py` too.)

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+  //Final step is generating a VFR video that has the proper offsets for each frame.
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: nit: space after `//`, expand VFR (like above).

File: lib/video/screenRecording/desktop/firefoxrecorder.js [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+  let stream = fs.createWriteStream(path.join(directoryName,'offsets.txt'));
+  stream.once('open', function() {
```
Comment (ncalexan) [inline lib/video/screenRecording/desktop/firefoxrecorder.js]: I am not 100% familiar with Node's callback timing, but is it guaranteed that this finishes before we continue?  Maybe `await new Promise(resolve => ... stream... resolve())`?

File: lib/video/screenRecording/recorder.js [hunk 17]
```diff
@@ -2,11 +2,21 @@
...
+const FirefoxRecorder = require('./desktop/firefoxrecorder');
+const os = require('os');
 
-module.exports = function getRecorder(options) {
+module.exports = function getRecorder(options, browser) {
   if (isAndroidConfigured(options)) {
     return new AndroidRecorder(options);
   } else {
-    return new X11Recorder(options);
+    if (
```
Comment (ncalexan) [inline lib/video/screenRecording/recorder.js:12]: Let's not block based on OS, since we expect only experts to use this.  And this doesn't agree with the docstring (right?), so let's update that too.

**Review by ncalexan — CHANGES_REQUESTED** [review]
Lots of nits, but nothing really significant.  I'd like to take another look before committing.  brennie is going to test this on Linux; I'll take a look on Android.  Great work!

**Comment by ncalexan** [comment]
My testing with Barret revealed a couple more issues:
1.  the `del` dependency needs to be added; see https://github.com/ncalexan/browsertime/commit/ae90cff5b2057b3820b3019b803456437954567f#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R12
2. we blat `windowrecorder-...` into the current working directory.  This is Firefox itself, right?  Can we get these put somewhere out of the way?
3. can we use the method [here](https://video.stackexchange.com/a/20590) to avoid `mp4fpsmod`?
4. the file is under `desktop` but isn't Desktop specific.  Lift it up to somewhere else in the hierarchy.

**Comment by dpalmeiro** [comment]
Thanks for the review!  I wasn't planning on uplifting this to sitespeedio until we get all of the final pieces resolved (like synchronous I/O) and it hits an actual release.  And some things will have to change for that uplift (like disabling this by default,  and we might also consider adding windows gdi/dshow screen capture support too).  

>     1. the `del` dependency needs to be added; see [ncalexan@ae90cff#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R12](https://github.com/ncalexan/browsertime/commit/ae90cff5b2057b3820b3019b803456437954567f#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R12)

Yup, will add that in.

>     2. we blat `windowrecorder-...` into the current working directory.  This is Firefox itself, right?  Can we get these put somewhere out of the way?

Yes,  this is firefox.  However, we can use the pref Bas added in to redirect the output images to somewhere inside of browsertime-results.  It will still create the windowrecorder-* directories, however, but that should be fine.

>     3. can we use the method [here](https://video.stackexchange.com/a/20590) to avoid `mp4fpsmod`?

Thanks, will try it out.  

>     4. the file is under `desktop` but isn't Desktop specific.  Lift it up to somewhere else in the hierarchy.

Hmm,  makes sense but not sure where to put it.  Maybe just sitting under "screenRecording", although that's a bit of an awkward location.   Could maybe just create a subdirectory under screenRecording for firefox.

File: lib/core/engine/collector.js (hunk 1 above)
Comment (dpalmeiro) [reply lib/core/engine/collector.js]: This isn't something we would uplift.  I think it's just more valuable for our purposes to display VC85 instead of PerceptualSpeedIndex on each iteration.  That being said, you're right and it probably should be in another commit.

File: lib/core/engine/command/measure.js (hunk 2 above)
Comment (dpalmeiro) [reply lib/core/engine/command/measure.js:66]: Unfortunately setting the background orange will insert a div element, so doing this twice inserts 2 elements we need to remove later on and that can add unnecessary delay.  Not sure if there's any non-awkward solution for this unless we can have Firefox emit the current frame on recording start.

File: lib/support/cli.js (hunk 4 above)
Comment (dpalmeiro) [reply lib/support/cli.js:239]: For the mozilla fork, I think this should be enabled by default.  We have evidence that screen capture overhead can be quite significant and on Windows there isn't even any support for it at the moment.   Will probably need to flip it to false for an uplift to sitespeedio though.  Let me know what you think.

File: lib/video/screenRecording/recorder.js (hunk 17 above)
Comment (dpalmeiro) [reply lib/video/screenRecording/recorder.js:12]: Blocking on OS for now until the other compositors are ready, but it might be ready now from Barret's patch so I'll have to test that out.  However, I do think this should be the default recorder over ffmpeg screen capture, especially for mach command usage.

File: lib/video/screenRecording/desktop/firefoxrecorder.js (hunk 16 above)
Comment (dpalmeiro) [reply lib/video/screenRecording/desktop/firefoxrecorder.js]: Yes, this is a good idea.

**Comment by ncalexan** [comment]
Rebased onto v4.9.3, minor touchups, and landed: https://github.com/mozilla/browsertime/commit/0b5aecd4cfa5dcc90773be14edbbe5ddbdeb2a74
